zephyr: Upgrade to Zephyr v4.0.0.

MaureenHelm opened the pull request:
<!-- Thanks for submitting a Pull Request! We appreciate you spending the
     time to improve MicroPython. Please provide enough information so that
     others can review your Pull Request.

     Before submitting, please read:
     https://github.com/micropython/micropython/blob/master/CODEOFCONDUCT.md
     https://github.com/micropython/micropython/wiki/ContributorGuidelines

     Please check any CI failures that appear after your Pull Request is opened.
-->

### Summary

<!-- Explain the reason for making this change. What problem does the pull request
     solve, or what improvement does it add? Add links if relevant. -->

Updates the Zephyr port to the latest Zephyr release tag. ~~It's currently a release candidate, not a final release, therefore this PR is in draft status until the Zephyr release is finalized.~~ [Zephyr v4.0.0](https://github.com/zephyrproject-rtos/zephyr/releases/tag/v4.0.0) was released on 15 Nov 2024.

### Testing

<!-- Explain what testing you did, and on which boards/ports. If there are
     boards or ports that you couldn't test, please mention this here as well.

     If you leave this empty then your Pull Request may be closed. -->

Tested on max32690fthr and frdm_k64f boards.

MaureenHelm committed:
zephyr: Change thread_analyze() to pass a cpu argument.

Zephyr v4.0.0 changed thread analyzer functions to pass a cpu argument
and allow thread analysis on each cpu separately. Promote this change to
the MicroPython zephyr module.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
tools/ci.sh: Upgrade Zephyr docker image 0.27.4 and SDK 0.17.0.

Upgrades CI to use the latest versions of the Zephyr docker image and
Zephyr SDK.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
zephyr: Upgrade to Zephyr v4.0.0.

Updates the Zephyr port build instructions and CI to use the latest
Zephyr release tag.

Tested on max32690fthr and frdm_k64f.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

dpgeorge commented:
The CI is failing with:
```
/micropython/ports/zephyr/main.c:100:50: error: 'CONFIG_SDMMC_VOLUME_NAME' undeclared
```

MaureenHelm committed:
zephyr: Remove CONFIG_SDMMC_VOLUME_NAME.

Zephyr v4.0.0 removed CONFIG_SDMMC_VOLUME_NAME and set the disk name
based on a devicetree property to allow multiple instances of the
sdmmc_subsys disk driver to be registered. All in-tree zephyr
definitions of the devicetree property are set to "SD", therefore we use
this constant string when initializing the MicroPython VFS.

Unfortunately I don't have hardware to verify that this change works
correctly at runtime; I can only verify that it resolves a build error
found when upgrading to MicroPython to Zephyr v4.0.0.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm commented:
@dpgeorge not sure what's wrong with the windows builds or why this PR would cause them to fail. can you take a look?

dpgeorge commented:
Merged as part of #16879 in commit 3c8d1b13f5b7e60e491e3a19734c237e1485e413
